pandas/tests/series/test_analytics.py

```diff
@@ -237,6 +238,60 @@ def test_npdiff(self):
         r = np.diff(s)
         assert_series_equal(Series([nan, 0, 0, 0, nan]), r)
 
+    def test_diff(self):
+        # Combined datetime diff, normal diff and boolean diff test
+        ts = tm.makeTimeSeries(name="ts")
+        ts.diff()
+
+        # int dtype
+        a = 10000000000000000
+        b = a + 1
+        s = Series([a, b])
+
+        result = s.diff()
+        assert result[1] == 1
+
+        # neg n
+        result = ts.diff(-1)
+        expected = ts - ts.shift(-1)
+        assert_series_equal(result, expected)
+
+        # 0
+        result = ts.diff(0)
+        expected = ts - ts
+        assert_series_equal(result, expected)
+
+        # datetime diff (GH3100)
+        s = Series(date_range("20130102", periods=5))
+        result = s - s.shift(1)
+        expected = s.diff()
+        assert_series_equal(result, expected)
+
+        # timedelta diff
+        result = result - result.shift(1)  # previous result
+        expected = expected.diff()  # previously expected
+        assert_series_equal(result, expected)
+
+        # with tz
+        s = Series(
+            date_range("2000-01-01 09:00:00", periods=5, tz="US/Eastern"), name="foo"
+        )
+        result = s.diff()
+        expected = Series(TimedeltaIndex(["NaT"] + ["1 days"] * 4), name="foo")
+        assert_series_equal(result, expected)
+
+        # boolean series (test for fixing #17294)
+        s = Series([False, True, True, False, False])
+        result = s.diff()
+        expected = Series([nan, True, False, True, False])
+        assert_series_equal(result, expected)
+
+        # boolean nan series
+        s = Series([False, True, nan, False, False])
+        result = s.diff()
```

Comment:
this routine is way too long can you parameterize

Reply:
Come on people, make up your minds. I have two versions now, the original commit is the original implementation, but working with a new numpy version, the second one also dealing with np.nan object 'boolean' arrays. 

Which one should be implemented? 
@TomAugspurger and @WillAyd want the second implementation
@jreback and me the first
That leaves @jbrockmendel and @simonjayhawkins

Reply:
@jreback Shall I rewrite the rest of those tests in pandas as well? I did not write the tests the way they are, I only added a few lines... I would rather do this in a refactor.

Reply:
Is there a particular reason why you moved this test? I think would be easier just to create another test specific to bool here and leave the time series stuff where it is